### PR TITLE
Moved position of card editor toggle button to bottom left

### LINKED ISSUE #4086: Move position of card editor toggle button to bottom left
If I recall correctly, the button to toggle between code and UI editor originally used to be in the bottom left corner. I think we should consider moving it back down there.

editor-toggle

**Benefits**:
* It makes all the card editors look significantly cleaner IMO: 
  * Lines up with the other cancel/save buttons
  * Always in a consistent spot rather than having a random button in the middle
  * Saves a little screen real estate in some cases, see the red rectangle in my gauge card example above
  * Button would now be visible at all times. The current placement of the button means that if a scrollbar becomes visible, the button becomes "invisible" and must be scrolled into view ([pic](https://i.imgur.com/k7LBGEe.jpg)) 
* Better mobile editing experience
  * Prevents accidental taps, because the button will be out of the way when scrolling. Right now you have to scroll past this button in order to see the preview (edit the "Entertainment" card in the demo on a mobile device to see what I mean). It is possible to accidentally enter the code editor, which I suspect most people do not want to do on a phone.
  * Moving the button will allow you to see a little more of the preview on mobile without scrolling

**Drawbacks**:
There is only one drawback to this as far as I can tell, which is demonstrated in the screenshot comparison below. Top screenshot is the current behavior and bottom screenshot is what it would look like under this proposal.

drawback

Basically when the preview is significantly taller than the editor area, there can be a noticeable gap before you see the toggle button. This scenario is pretty rare and mostly only occurs when you switch to the Code editor on cards with little to no configuration options. The above screenshot is the most extreme example I could create using the demo.

Visually this gap is not ideal, but I personally don't consider it a dealbreaker, especially when taking into account the other benefits. Curious what others think, is this acceptable or not? I feel like in these instances where there's such limited configuration options, the user wouldn't really be toggling the editor back and forth anyway, so they wouldn't notice this.

Here is how the same Thermostat card looks in visual editor with this change: 

thermostat-card-visual-editor

It doesn't have the huge gap that the code editor does, and the user would have no reason to switch to code editor anyway since all the options are present here. So maybe the drawback mentioned above isn't such a big deal? I dunno.

Hopefully I did a decent job demonstrating the pros and cons.

### PR DESCRIPTION
**Pull request by springstan**
Fixes: #4086

Description:
- moved card editor toggle buton from hui-card-editor.ts to hui-dialog-edit-card.ts
- removed editor toggle button css in hui-card-editor.ts
- added getter for _warning and _GUImode in hui-card-editor.ts
- added css to place editor toggle button to the bottom left in hui-dialog-edit-card.ts
- added helper functions in hui-dialog-edit-card.ts to toggle the _guiMode and get _warning and _error from hui-card-editor.ts

Results:
code
visual

### CODE REVIEW

File: src/panels/lovelace/editor/card-editor/hui-dialog-edit-card.ts [hunk 1]
```diff
@@ -44,6 +44,8 @@ export class HuiDialogEditCard extends LitElement {
   @property() private _saving: boolean = false;
   @property() private _error?: string;
 
+  @property() private _GUImode: boolean = true;
```
Comment (bramkragten) [inline src/panels/lovelace/editor/card-editor/hui-dialog-edit-card.ts]: What happens if a card doesn't have a GUI?

File: src/panels/lovelace/editor/card-editor/hui-dialog-edit-card.ts [hunk 2]
```diff
@@ -136,6 +153,18 @@ export class HuiDialogEditCard extends LitElement {
     `;
   }
 
+  private _isDisabled(): boolean {
```
Comment (bramkragten) [inline src/panels/lovelace/editor/card-editor/hui-dialog-edit-card.ts]: This should be a getter

**Comment by bramkragten** [comment]
What is making sure the button re-renders when the state of the editor changes?

File: src/panels/lovelace/editor/card-editor/hui-dialog-edit-card.ts (hunk 1 above)
Comment (springstan) [reply src/panels/lovelace/editor/card-editor/hui-dialog-edit-card.ts]: Good point, how can I check that?

**Comment by springstan** [comment]
I have tested it and so far, the innerHTML and disabled state updates automatically.
Or do I need to implement an updated function like in hui-card-editor.ts?

File: src/panels/lovelace/editor/card-editor/hui-dialog-edit-card.ts (hunk 1 above)
Comment (bramkragten) [reply src/panels/lovelace/editor/card-editor/hui-dialog-edit-card.ts]: Edit a stack

Comment (springstan) [reply src/panels/lovelace/editor/card-editor/hui-dialog-edit-card.ts]: Sorry what do you mean?

File: src/panels/lovelace/editor/card-editor/hui-dialog-edit-card.ts (hunk 2 above)
Comment (springstan) [reply src/panels/lovelace/editor/card-editor/hui-dialog-edit-card.ts]: Fixed in https://github.com/home-assistant/home-assistant-polymer/pull/4112/commits/1aa9c5800abb6f68b3d3f5505d41604c1dfd9fb6

File: src/panels/lovelace/editor/card-editor/hui-dialog-edit-card.ts (hunk 1 above)
Comment (bramkragten) [reply src/panels/lovelace/editor/card-editor/hui-dialog-edit-card.ts]: No, but it seems to give a warning and we don't show the toggle button when there is a warning right?

Comment (springstan) [reply src/panels/lovelace/editor/card-editor/hui-dialog-edit-card.ts]: Yeah you are right, but how can I check for this warning if the custom element that detects it is not rendered yet?

Comment (bramkragten) [reply src/panels/lovelace/editor/card-editor/hui-dialog-edit-card.ts]: `hui-card-editor` should always be in control about the GUI state, not the dialog

**Comment by bramkragten** [comment]
> I have added the updated function to make sure that the button re-renders

That is not how it works. Updated is called after an update.
Properties trigger updates. On a change of a property an update is called and the element re-renders.
Check https://lit-element.polymer-project.org/guide/lifecycle

You should remove the `_GUImode` property from `hui-dialog-edit-card` and instead listen to updates of  this._cardEditorEl.GUImode. If that changes `hui-dialog-edit-card` should re-render.

**Comment by springstan** [comment]
I will take a look at that and try to implement it, thanks :)

**Comment by springstan** [comment]
I looked through the documentation and found `someProperty.hasChanged`, however I could not figure out how to use it from another element or listen to any state changes.
Still the issue with cards which have no GUI Editor persists, not really sure how to handle that.

**Comment by bramkragten** [comment]
Fire an event from the child element that the parent listens for. That way you always know when it changes.

**Comment by springstan** [comment]
So fire the event from `hui-card-editor` and listen with `hui-dialog-edit-card`? In addition, where should I listen for this event? Maybe in the render function?

**Comment by bramkragten** [comment]
Event listeners can be added in render to a component with `@event-name=${functionToCallToHandleTheEvent}`

**Comment by springstan** [comment]
Sorry I do not know how to implement this. Furthermore, I could not find an example in this codebase that uses `@event-name=${functionToCallToHandleTheEvent}`. 
Could you please provide me some details? Thanks in advance :)

**Comment by bramkragten** [comment]
Here is an example:
https://github.com/home-assistant/home-assistant-polymer/pull/4112/files#diff-50629a7d49f83e694808066b16545d74R138

It is used a lot in our code base :-) It is the event listener syntax of `lit-element`

**Comment by springstan** [comment]
@bramkragten oh yes you are right, I completely misread `@event-name=${functionToCallToHandleTheEvent}`, sorry my bad. I will try to work it out now :')

**Comment by bramkragten** [comment]
> @bramkragten oh yes you are right, I completely misread `@event-name=${functionToCallToHandleTheEvent}`, sorry my bad. I will try to work it out now :')

No problem, might have been me not explaining it well. If you have have questions, you can always drop them in the discord frontend dev chat.

**Comment by springstan** [comment]
Whoops completely forgot about this PR 😬
Tried to implement firing and listening to the event, firing works but listening does not.
@bramkragten could you help me so this can finally be completed? 😅

**Comment by bramkragten** [comment]
You are listening on a button, you should be listening on a parent of the element that fires the event.

**Comment by springstan** [comment]
So somewhere along these lines?
https://github.com/home-assistant/home-assistant-polymer/blob/4dca3289f60b6fc69cf1442cb49c74a1f3faec25/src/panels/lovelace/editor/card-editor/hui-dialog-edit-card.ts#L103-L107

**Comment by bramkragten** [comment]
Yes

**Comment by bramkragten** [comment]
@springstan are you going to finish this or should we close it?

**Comment by springstan** [comment]
I am really busy at the moment, therefore I will close this PR.

**Pull request closed without merging**